executor/insert_common.go

```diff
@@ -252,6 +252,7 @@ func (e *InsertValues) evalRow(ctx context.Context, list []expression.Expression
 			return nil, err
 		}
 		val1, err := table.CastValue(e.ctx, val, e.insertColumns[i].ToInfo())
+		e.handleOverflowWarning(e.insertColumns[i], rowIdx)
```

Comment:
If we want to insert N rows, is the complexity of this operation for this SQL `O(N*N)`?
And why do we have to do a converting?
How about letting the underlying expression return error `ErrWarnDataOutOfRange` directly?

Reply:
Return `ErrWarnDataOutOfRange` directly is not compatible with MySQL.
As MySQL return `Out of range value for column 'a' at row 1`,
but we return `constant -1 overflows bigint`,
PR([#9829](https://github.com/pingcap/tidb/pull/9829)) does not fix this problem. 

Another way to do this is modify `func (sc *StatementContext) HandleTruncate(err error) error`,
delete `TruncateAsWarning`, and do this in upper function, but this will introduce some repetitive code.

Reply:
> Return `ErrWarnDataOutOfRange` directly is not compatible with MySQL.
> As MySQL return `Out of range value for column 'a' at row 1`,
> but we return `constant -1 overflows bigint`,
> PR([#9829](https://github.com/pingcap/tidb/pull/9829)) does not fix this problem.
> 
> Another way to do this is modify `func (sc *StatementContext) HandleTruncate(err error) error`,
> delete `TruncateAsWarning`, and do this in upper function, but this will introduce some repetitive code.

No, after my pr, TiDB return `ERROR 1264 (22003): Out of range value for column 'a' at row 1`

Reply:
> Return `ErrWarnDataOutOfRange` directly is not compatible with MySQL.
> As MySQL return `Out of range value for column 'a' at row 1`,
> but we return `constant -1 overflows bigint`,
> PR([#9829](https://github.com/pingcap/tidb/pull/9829)) does not fix this problem.
> 
> Another way to do this is modify `func (sc *StatementContext) HandleTruncate(err error) error`,
> delete `TruncateAsWarning`, and do this in upper function, but this will introduce some repetitive code.

As I pointed out in my pr, the root cause of the problem is that TiDB uses the same logic to deal with insert and cast(). And I had tested your code,  your codes doesn't work correctly with `insert into t values (0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF);` and `insert into t values (0x8000000000000001);`
```
mysql> select tidb_version();
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                                              |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v3.0.0-beta-936-g72b42f6ee
Git Commit Hash: 72b42f6eed76f9aab08b574a915963b27786d223
Git Branch: fix-hex-values
UTC Build Time: 2019-08-05 09:23:25
GoVersion: go version go1.12.4 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> set sql_mode="";
Query OK, 0 rows affected (0.00 sec)

mysql> create table t(c tinyint);
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t values (0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF);
Query OK, 1 row affected, 1 warning (0.00 sec)

mysql> insert into t values (0x8000000000000001);
Query OK, 1 row affected (0.00 sec)

mysql> select * from t;
+----------------------+
| c                    |
+----------------------+
|                   -1 |
| -9223372036854775807 |
+----------------------+
2 rows in set (0.00 sec)
```

Reply:
@wjhuang2016 
Thank you for pointing out the error test case in my PR, I will fix it.
But PR(#9829) Warning Message is not compatible with MySQL
```
mysql> select tidb_version();
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                                               |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v3.0.0-beta-926-gc69cdf072
Git Commit Hash: c69cdf072100801f3d18dc9e7477a23f7f005e94
Git Branch: fix_hex_insert
UTC Build Time: 2019-08-06 03:25:31
GoVersion: go version go1.12.6 darwin/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> set sql_mode="";
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t values (0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF);
Query OK, 1 row affected, 2 warnings (0.02 sec)

mysql> show warnings;
+---------+------+------------------------------------------------------------------------+
| Level   | Code | Message                                                                |
+---------+------+------------------------------------------------------------------------+
| Warning | 1292 | Truncated incorrect BINARY value: '0xffffffffffffffffffffffffffffffff' |
| Warning | 1690 | constant 18446744073709551615 overflows tinyint                        |
+---------+------+------------------------------------------------------------------------+
2 rows in set (0.00 sec)

mysql> insert into t values (0x8000000000000001);
Query OK, 1 row affected, 1 warning (0.01 sec)

mysql> show warnings;
+---------+------+------------------------------------------------+
| Level   | Code | Message                                        |
+---------+------+------------------------------------------------+
| Warning | 1690 | constant 9223372036854775809 overflows tinyint |
+---------+------+------------------------------------------------+
1 row in set (0.00 sec)

mysql> insert into t values (0xFFFFFFFFFFF);
Query OK, 1 row affected, 1 warning (0.01 sec)

mysql> show warnings;
+---------+------+-------------------------------------------+
| Level   | Code | Message                                   |
+---------+------+-------------------------------------------+
| Warning | 1690 | constant 17592186044415 overflows tinyint |
+---------+------+-------------------------------------------+
1 row in set (0.01 sec)

```

Reply:
> @wjhuang2016
> Thank you for pointing out the error test case in my PR, I will fix it.
> But PR(#9829) Warning Message is not compatible with MySQL
> 
> ```
> mysql> select tidb_version();
> +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | tidb_version()                                                                                                                                                                                                                                                                                                                               |
> +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | Release Version: v3.0.0-beta-926-gc69cdf072
> Git Commit Hash: c69cdf072100801f3d18dc9e7477a23f7f005e94
> Git Branch: fix_hex_insert
> UTC Build Time: 2019-08-06 03:25:31
> GoVersion: go version go1.12.6 darwin/amd64
> Race Enabled: false
> TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
> Check Table Before Drop: false |
> +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> 1 row in set (0.00 sec)
> 
> mysql> set sql_mode="";
> Query OK, 0 rows affected (0.00 sec)
> 
> mysql> insert into t values (0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF);
> Query OK, 1 row affected, 2 warnings (0.02 sec)
> 
> mysql> show warnings;
> +---------+------+------------------------------------------------------------------------+
> | Level   | Code | Message                                                                |
> +---------+------+------------------------------------------------------------------------+
> | Warning | 1292 | Truncated incorrect BINARY value: '0xffffffffffffffffffffffffffffffff' |
> | Warning | 1690 | constant 18446744073709551615 overflows tinyint                        |
> +---------+------+------------------------------------------------------------------------+
> 2 rows in set (0.00 sec)
> 
> mysql> insert into t values (0x8000000000000001);
> Query OK, 1 row affected, 1 warning (0.01 sec)
> 
> mysql> show warnings;
> +---------+------+------------------------------------------------+
> | Level   | Code | Message                                        |
> +---------+------+------------------------------------------------+
> | Warning | 1690 | constant 9223372036854775809 overflows tinyint |
> +---------+------+------------------------------------------------+
> 1 row in set (0.00 sec)
> 
> mysql> insert into t values (0xFFFFFFFFFFF);
> Query OK, 1 row affected, 1 warning (0.01 sec)
> 
> mysql> show warnings;
> +---------+------+-------------------------------------------+
> | Level   | Code | Message                                   |
> +---------+------+-------------------------------------------+
> | Warning | 1690 | constant 17592186044415 overflows tinyint |
> +---------+------+-------------------------------------------+
> 1 row in set (0.01 sec)
> ```

Yeah, The warings here is very tricky, and I think convert it here is the best to do.

Reply:
So after discussing, is this PR more appropriate than #9829?
What can we do on #9829 and if it's nothing to do, can we close it? @wjhuang2016